### PR TITLE
Feat(code): Additional Zoom Levels & Zoom smoothing

### PR DESCRIPTION
**Pull request by Zitchas**
**Feature:** This PR implements an interim fix to help with the situation described in #8071 

## Feature Details
1. Adds in a couple steps on the higher end of the zoom scale.
2. Adds in a couple steps on the low end of the spectrum
3. Re-defines the set of zooms to provide a lot more granularity and smoothness

Formerly the farthest out zoom was .25. Now it is 0.105

Formerly the closest in zoom was 2. Now it is 3.364

Previously the zoom changed by leaps (doubling the area with each step lower in zoom)
Now it increases by 41.14%. I don't know what that's special, but warpcore likes the zoom changing by the fourth root of 2.

This makes it slightly slower to transition from min to max a little, but it makes it look much smoother and much easier to find the perfect zoom to match the situation.

This also adds commentary so that people learning the code can identify this section and understand what it is doing.

As an aside, for people working out the math (say, in libreoffice Calc or Excel)
1. Start with a value of 1 (for no zoom)
2. To get more detail, multiply that by 2^(1/4) for the next level of zoom.
3. Repeat until the desired maximum level of zoom.
4. To see more of the system, divide 1 by 2^(1/4) for the next level of zoom.
5. Repeat until the desired maximum area of the system is visible.

## UI Screenshots
Not a change to the UI itself, just the range of options for the main action area.

## Testing Done
Played the game, tried out playing at both max and min values.

### Automated Tests Added
N/A

## Performance Impact
- At the maximum zoom the game is basically displaying the entire system, which doesn't seem to have a significant impact, but I suppose it might.

### CODE REVIEW

File: source/Preferences.cpp [hunk 1]
```diff
@@ -37,7 +37,10 @@ namespace {
 	const string EXPEND_AMMO = "Escorts expend ammo";
 	const string FRUGAL_ESCORTS = "Escorts use ammo frugally";
 
-	const vector<double> ZOOMS = {.25, .35, .50, .70, 1.00, 1.40, 2.00};
+	// This controls the range of zoom scales the player can switch to.
+	// larger values are close-up and small values are farther away
+	const vector<double> ZOOMS = {.1, .2, .3, .4, .5, .6, .7, .8, .9, 1.0,
+		1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 3.0, 4.0};
 	int zoomIndex = 4;
```
Comment (warp-core) [inline source/Preferences.cpp:44]: The default zoom is now 0.5.

Comment (Zitchas) [reply source/Preferences.cpp:44]: So that zoomIndex is specifying the default zoom level? So shifting it up to 7 would keep the default zoom the same.

Comment (warp-core) [inline source/Preferences.cpp:44]: > Previously the zoom progressed erratically: .25, .35, .5, .7, .1, .1.4, 2

The previous scale was not actually erratic at all. It went up in factors of approximately the square root of two with each step.
In game, this means that moving by two steps results in the dimensions doubling (or halving).

Your version, on the other hand, will appear very erratic because the factor by which the zoom level changes with each step also changes with each step, and whatever pattern is there is copmletely discarded at the very highest zoom levels.

```suggestion
	const vector<double> ZOOMS = {.125, 1.75, .25, .35, .5, .7, 1., 1.4, 2., 2.8, 4.};
	int zoomIndex = 6;
```

There's also the issue that existing preference files will now not refer to the same zoom value. That can be remedied by putting all new valuse at the end of the vector.

```suggestion
	const vector<double> ZOOMS = {.25, .35, .5, .7, 1., 1.4, 2., 2.8, 4., .125, .175};
	int zoomIndex = 4;
```

Comment (warp-core) [reply source/Preferences.cpp:44]: No, the index of `1.` is now 9.

Comment (Zitchas) [reply source/Preferences.cpp:44]: Where does it specify the index? And how does "1." point to 9?

**Comment by RisingLeaf** [comment]
Can you add a screen recording?

File: source/Preferences.cpp (hunk 1 above)
Comment (warp-core) [reply source/Preferences.cpp]: Indexing is a property of the programming language.

https://www.cpp.edu/~elab/ECE114/Array.html#:~:text=An%20array%20is%20a%20collection,each%20with%20its%20own%20identifier).

Here's an article that might help.

In programming, "index" is a number that denotes some specific element in a container.
In many programming languages, including C++, container indexing starts at 0, meaning the index of the first element of the container is 0.

The existing vector (`vector` is a specific type of container in C++, it is essentially a standard "array," but its size can be changed dynamically, a standard C++ array does not support this) has these elements at the following indices:

index | 0 | 1 | 2 | 3 | 4 | 5
-- | -- | -- | -- | -- | -- | --
element | .25 | .35 | .5 | 1. | 1.4 | 2.

Your version has:

index | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
element | .1 | .2 | .3 | .4 | .5 | .6 | .7 | .8 | .9 | 1. | 1.1 | 1.2 | 1.3 | 1.4 | 1.5 | 1.6 | 1.7 | 1.8 | 1.9 | 2. | 3. | 4.

**Comment by Zitchas** [comment]
Unfortunately I can't. I used to use my graphic driver's built in screen capture, but those options and ability to do so seem to have vanished, so I currently do not have a means of doing so.

File: source/Preferences.cpp (hunk 1 above)
Comment (Zitchas) [reply source/Preferences.cpp]: Ah, yes, I was forgetting that it started at 0, not 1. That makes more sense.

**Comment by RisingLeaf** [comment]
https://user-images.githubusercontent.com/85687254/211221150-0388076e-184d-4af6-8937-a92402f17e30.mov

Did the work for you XD

**Comment by Zitchas** [comment]
Could you explain the math/formula behind the scale values? The intent behind switching to the .1, .2, etc was to have shorter gaps between zoom levels, because I often find myself jumping back and forth between levels, unhappy with both.

**Comment by RisingLeaf** [comment]
> Screen.Recording.2023-01-08.at.1.59.40.PM.mov
> 
> Did the work for you XD

It seems like it gets crowded, of course this is a plugin, but we may want to look at ringworlds

**Comment by warp-core** [comment]
> Could you explain the math/formula behind the scale values? The intent behind switching to the .1, .2, etc was to have shorter gaps between zoom levels, because I often find myself jumping back and forth between levels, unhappy with both.

The scale goes up in factors of √(2) (approximately 1.4), meaning each step *always* halves the area that is visible.

**Comment by Saugia** [comment]
I'm concerned about adding more deeper zoom when it comes to artwork. All ships were made under the premise of the current zoom, and I would not like for us to have to redo all of them *again* to match this new depth. A lot of things you can't see normally, awkward things or design choices that make sense at the current limit are quickly all thrown out the window.

So I'm a bit iffy on the increased zoom, but not so much the farther out one.

**Comment by Zitchas** [comment]
> I'm concerned about adding more deeper zoom when it comes to artwork. All ships were made under the premise of the current zoom, and I would not like for us to have to redo all of them _again_ to match this new depth. A lot of things you can't see normally, awkward things or design choices that make sense at the current limit are quickly all thrown out the window, and will be more apparent.

Blame your own success. :) Part of the reason I want a closer zoom is specifically so I can look at my flagship in greater detail. That, and watch the turrets spin. There's really not much other reason for making my Osprey take up a quarter of my screen (top to bottom, not area)

**Comment by Zitchas** [comment]
> > Could you explain the math/formula behind the scale values? The intent behind switching to the .1, .2, etc was to have shorter gaps between zoom levels, because I often find myself jumping back and forth between levels, unhappy with both.
> 
> The scale goes up in factors of √(2) (approximately 1.4), meaning each step _always_ halves the area that is visible.

I think I probably need to go get a math refresher, because I'm not seeing the pattern. From the first step to the second step is an increase of .03, and then from there to the third step is +0.05, then +0.1, then +.15, then +.2, then +.3, the +.4, then +.6, then +.8

Well, I'm seeing what looks like it might be a pattern with some digits rounded off, but I'm not seeing how those increases are connected to √(2). To me it looks like +0.05 three times, then +0.1 twice, then +0.2 twice.

**Comment by Zitchas** [comment]
Alright, now that I understand how the zoom is calculated, I think this set should achieve the smoother and smaller steps that I was going for, but do it without the irregular aspects that the old set was doing.

Each step up the sequence increases the area by +50% instead of +100%.

**Comment by 1010todd** [comment]
> Blame your own success. :) Part of the reason I want a closer zoom is specifically so I can look at my flagship in greater detail. That, and watch the turrets spin. There's really not much other reason for making my Osprey take up a quarter of my screen (top to bottom, not area)

Another art problem is we might hit a limit of how optimized a piece can be for various level of zoom, if we make it very detailed and looked good at max zoom in it might turn out to be garbled mess at further zoom out. Also at current max zoom in I can already notice that Shuttles are blurry, sure High DPI plugin helps with that but it's not the standard (and I'm not sure if we want that). Since you mentioned that it doesn't really have a functional reason other than viewing ship details I think it could be unlocked by High DPI plugin instead.

**Comment by EjoThims** [comment]
Haven't had a chance to check this out in game myself, but in concept I'm thinking that including the base code but only utilizing it via the high DPI plug-in does indeed sound like it may be the best path.

This would, of course, also allow the use of another plug-in to activate it without including the high DPI images, if someone really did want that option.

**Comment by Zitchas** [comment]
Hmmm... That's true. I am using the high-dpi plug-in; and honestly, for testing/review purposes it's really nice to be able to see *exactly* where the hardpoints are.

For instance, I've noticed that the shots from the particle cannons on my Osprey actually start at the base of the gun on the Osprey  instead of on the tip of the protruding "guns"...

**Comment by EjoThims** [comment]
> Can the high-dpi unlock additional zoom levels? Or is that a "more coding required" sort of thing?

If it's merged as a setting that's defaulted to "off," plug-in should be able to then set it to "on," right?

**Comment by Zitchas** [comment]
I didn't think plug-ins could touch anything inside the source folder, nor touch anything in the user settings, either.

**Comment by 1010todd** [comment]
Maybe when gamerule is merged it could use that.

**Comment by quyykk** [comment]
I definitely agree that you shouldn't be able to zoom in too much. On the max zoom level it starts to get blurry a bit (without the high dpi plugin).

I would also say that you shouldn't be able to zoom out that much either, because it is possible to see the haze (and asteroids) repeating, which isn't ideal.

**Comment by 1010todd** [comment]
Repeating asteroid is already noticeable at current max zoom out anyway. Haze is noticeable when you fly in one direction really fast but it takes a while before you actually notice it so I think it might be fine. Else maybe we could use multiple haze to add some variation.

For asteroid, just remove it and put all of them in a belt. People seems to hate the tiled asteroids anyway (Disclaimer: I haven't done actual poll yet so might be wrong.). All tiled asteroid did is occasionally be annoying to ranged weapons and missiles and doesn't really provide any interesting gameplay other than "Oh range sucks in this system so I won't fight here.", a proper belt provides actual terrain player can maneuver around to avoid missiles, you know, actually fly your ship not just go forward and fire.

**Comment by Zitchas** [comment]
On the more long term side of things, I don't think we should be scaling the haze on such a 1:1 basis. Most haze looks like nebulae gases in the far distance, Likewise, stars. Both haze and stars should get scaled into the background at a *much* slower rate than all the ships and stuff that is on our plane of action. Stuff that's many lightyears away really shouldn't be scaling much at all when our camera is moving mere light-seconds or minutes. 

I'll reduce the scope of the additional zoom levels on this, but longer term I'd like to see some better zoom handling of haze and stars in particular; and perhaps a shift to a tactical radar-type display when our main zoom exceeds a certain level (among other changes, the tiled asteroids would probably be just gone entirely on such a display, and mineables would probably be represented by bands instead of discrete individuals.) But that's probably something for an issue/request rather than this PR.

**Comment by Hurleveur** [comment]
What about adding a preference, "extended zooms" that, if on, allows going further than the current in game zoom?
Not necessarily for both in and out

**Comment by Hurleveur** [comment]
Max and min zoom would be sound as a gamerule I think.

**Comment by quyykk** [comment]
> Max and min zoom would be sound as a gamerule I think.

sounds like something that belongs in the interface file, not a gamerule (it's not a game mechanic)

**Comment by Zitchas** [comment]
Amusingly, I didn't use the square root of 1.5, although it is entirely possible that's the same value I used. I made a spreadsheet, and calculated the areas of a circle for the baseline 1 zoom (treating the zoom as being the radius), then increased and decreased it by 50% and continued up and down the scale list doing that, then worked backwards using that as the area to find out what the radius (ei, zoom) would be. So each step very precisely is 50% more (or less, depending on the direction) area than the next step.

I think my main thought for keeping it where it is, is that on the longterm, I don't think anyone is particularly served by just having arbitrarily large expanses of visible space in our current viewscreen setup, but if it is accompanied by the previously mentioned changes to the way we display things, then it suddenly becomes *very* useful.

I'd envision a range of parameters:
- Zoom going all the way out to values several steps below 0.108 (which pushes the center-to-edge of screen distance to values significantly over 10k, depending on the screen.
- Haze and starscape don't scale at the same rate as the system, instead only scaling to maybe a quarter as much as the system. Perhaps even have a variable scaling effect, where at the most zoomed in it scales at about full rate with the zoom, but as one zooms out, that decreases to something closer to a 5% rate of scaling at the farthest zoom out point.
- tiled asteroids fade out as the zoom passes through 0.132.
- At around 0.108 all ship sprites are replaced with iconography, probably based on the radar symbols, but our viewscreen has more area, so we can afford to have actual symbols for each category of ship. I'd expect general categories to have a symbol (ex. Freighter, warship, transport, utility) with the size of the icon being tied to the mass of the ship, thus obviating the need to have different symbols for light, medium, and heavy distinctions within a category. These would be color coded using the radar colors for IFF.
- Selection area for clicking on a ship would be increased a bit so one doesn't need pinpoint accuracy to select a ship by clicking on it.
- Perhaps additional info could be added to the screen at this point. Some ideas: Having the habitable band highlighted, if one has an asteroid scanner add in the mineable asteroid orbital paths, Maybe a "distance to the center of the system" measurement, with a direction indicator, Maybe show the system arrival and departure rings (if any are present), planetary orbits, etc. 
- Probably any other information that'd we want to put into the radar or the system-map from the map panel, but don't have the space to put in those areas.
- directional arrows identifying (and naming) the hyperlinks to surrounding systems.

Basically this would transform our main viewscreen so that at high zooms it switches from being merely a "lol, all of Sol fits within ta quarter of my screen" thing, and instead is an actual useful tactical / information mode

**Comment by warp-core** [comment]
> Amusingly, I didn't use the square root of 1.5, although it is entirely possible that's the same value I used. I made a spreadsheet, and calculated the areas of a circle for the baseline 1 zoom (treating the zoom as being the radius), then increased and decreased it by 50% and continued up and down the scale list doing that, then worked backwards using that as the area to find out what the radius (ei, zoom) would be. So each step very precisely is 50% more (or less, depending on the direction) area than the next step.

The ratio of the radii of two circles is equal to the square root of the ratio of their areas.
In choosing 50% (1.5x) as your area factor, you implicitly chose the square root of 1.5 as the zoom scale factor.

The rest of your comment reads to me like you are saying the changes here are pointless if we aren't also going to make all the other changes you suggest, including seemingly modelling the most zoomed out levels after a strategy game?

**Comment by Zitchas** [comment]
Well, they're not actually pointless, they just have decreasing value if we're not considering ways of making that space more valuable somehow. That could be by making systems exponentially bigger (which is I think may be what Vitalchip has, given that they've mentioned having 40k view distances), or use it as a platform for providnig the player with more information that would be useful/beneficial to have, but doesn't really fit on the other places it'd be applicable for.

As far as that goes, I'm using these on my normal gameplay, and I'm finding it really nice to be able to just zoom out and watch the flickers of gunfire in the distance. Or just admire the depths of space. It'd be nicer if it was blacker, but it's still nice either way.

**Comment by Zitchas** [comment]
OK, major changes at this point:
- Switched to a sequence based on the fourth root as suggested by warpcore
- Removed the "4" level, so now one can only zoom in to 3.364 instead of 4. 


Hopefully that mitigates the fuzziness that some people commented on at the closest zoom level. 
I don't know why warpcore suggested using the fourth root, as they didn't provide an explanation as to why they didn't like the square root of 1.5, so I can't tell if changing it has resulted in the situation being improved or not. In any case, I have tested them, and they do provide a smooth and flexible range of zoom options, so I'm happy with them.

**Comment by RisingLeaf** [comment]
Currently we are rendering ship sprites at half of their size to have smooth looking rotations of them. This change would break that, so I have a hard no for more zoom levels in.
Read this for more info: https://github.com/endless-sky/endless-sky/wiki/CreatingShips#sprites

**Comment by RisingLeaf** [comment]
On that note I would also like to say that max zoom is a value defined by the engine and should be defined there.

**Comment by Zitchas** [comment]
> Currently we are rendering ship sprites at half of their size to have smooth looking rotations of them. This change would break that, so I have a hard no for more zoom levels in. Read this for more info: https://github.com/endless-sky/endless-sky/wiki/CreatingShips#sprites

Question: Have you actually *tried* the PR to see how it looks? These comments sound very theoretical, and not based on actual observation.

**Comment by Zitchas** [comment]
@RisingLeaf There, I trimmed off the two highest zoom level increases for you.

Having those zooms were very nice to have, especially when flying small ships, but if those are "hard no" then so be it. They're gone.

And yes, there's still 2.378 above the 2,  but I firmly believe that having at least a *slightly* higher zoom is very much a benefit to the game, and has zero impact on game quality or blurriness.

**Comment by RisingLeaf** [comment]
I did test it and the ship gets blurry when you rotate.

**Comment by Zitchas** [comment]
Thank you. I appreciate you taking the time to test it and stating the results.

Was that before or after my recent push that eliminated the two highest zoom levels? (I assume before, since those haven't finished checks yet)

File: source/Preferences.cpp [hunk 2]
```diff
@@ -37,8 +37,11 @@ namespace {
 	const string EXPEND_AMMO = "Escorts expend ammo";
 	const string FRUGAL_ESCORTS = "Escorts use ammo frugally";
 
-	const vector<double> ZOOMS = {.25, .35, .50, .70, 1.00, 1.40, 2.00};
-	int zoomIndex = 4;
+	// This controls the range of zoom scales the player can switch to.
+	// larger values are close-up and small values are farther away
+	const vector<double> ZOOMS = {.105, .125, .149, .177, .210, .250, 0.297, 0.354, 0.420, 0.500, 0.595, 0.707, 0.841,
+		1.000, 1.189, 1.414, 1.682, 2.000, 2.378};
+	int zoomIndex = 11;
```
Comment (warp-core) [inline source/Preferences.cpp:44]: This level of precision is unnecessary, and we don't need trailing zeroes here.
```suggestion
	const vector<double> ZOOMS = {.105, .125, .15, .175, .21, .25, 0.3, 0.35, 0.42, 0.5, 0.6, 0.7, 0.84,
		1., 1.2, 1.4, 1.68, 2., 2.4};
	int zoomIndex = 13;
```

**Review by warp-core — COMMENTED** [review]
I notice in game that the hull ring of the status overlay is divided into segments.
With the existing zoom values, the ring is always divided equally, but many of the new zoom values here result in the final segment being truncated.

**Comment by RisingLeaf** [comment]
Got it, will remember that.
I just added the higher zoom to my local branch.

I am sorry if "hard no" sounded too harsh, it's just an engine/sprite limitation we have here.

So if you really want closer zoom you will have to convince art people to rerender and re postprocess all sprites.

**Comment by Zitchas** [comment]
> I just added the higher zoom to my local branch.

Easy enough, yes. That doesn't, however, tell me if you tested with the addition of the fairly minor increase to 2.378, or with the previous iteration that went up to 3.364

File: source/Preferences.cpp (hunk 2 above)
Comment (Zitchas) [reply source/Preferences.cpp]: Sorry, I pushed the change to the max zoom before applying this. I have pushed a second commit that removes all the extraneous zeroes and changed the index to 13.

Thank you for the suggestion, though. Normally I apply these first.

**Comment by RisingLeaf** [comment]
> > I just added the higher zoom to my local branch.
> 
> 
> 
> Easy enough, yes. That doesn't, however, tell me if you tested with the addition of the fairly minor increase to 2.378, or with the previous iteration that went up to 3.364

2.4 looks fine as far as I can tell.
3.4 is blurry again.

I tested all of them

**Comment by Zitchas** [comment]
So, should I put the 2.378 back in? Or leave it out?

**Comment by RisingLeaf** [comment]
I dont really have an opinion on that as I think it is fine either way

**Review by Hurleveur — APPROVED** [review]
With the expanded AI capabilities, fights are going to be bigger
So I see no reason not to merge this so we can get a better view of whats going on
We can always develop it more later, by adding more customization but in the meantime we're not forcing anyone to unzoom more. I for one would like that possibility, though

**Comment by warp-core** [comment]
Not merging this (or anything else) until we update the workflows with whatever new hashes GitHub is using.

**Comment by Zitchas** [comment]
While I realize that the precision might not be strictly *necessary*, doesn't it provide at least some potential benefit to keep the changes exactly according to the specified ratio, while having zero downside? I don't see the point to rounding off the values at all. As best I can tell, leaving them with that extra bit of precision is a zero downside and a clear (albeit minor) benefit.

**Comment by warp-core** [comment]
> While I realize that the precision might not be strictly _necessary_, doesn't it provide at least some potential benefit to keep the changes exactly according to the specified ratio, while having zero downside? I don't see the point to rounding off the values at all. As best I can tell, leaving them with that extra bit of precision is a zero downside and a clear (albeit minor) benefit.

These values are close enough. It is not possible to be exact, so the line has to be drawn somewhere, and the previous versions of this resulted in the status overlay looking bad at many of the new and existing zoom levels, as I mentioned in https://github.com/endless-sky/endless-sky/pull/8081#pullrequestreview-1271438304.

**Comment by Zitchas** [comment]
OK, so what you're saying is that there *is* a benefit to rounding them off. Sounds good, then.
